spec/strings_file_validation_helper_spec.rb

```diff
@@ -0,0 +1,66 @@
+require 'spec_helper'
+
+module Fastlane
+  UI = FastlaneCore::UI unless Fastlane.const_defined?('UI')
+
+  module Helper
+    class StringsFileValidationHelper
+      # Inspects the given `.strings` file for duplicated keys.
+      #
+      # @param [String] file The path to the file to inspect.
+      def self.find_duplicated_keys(file:)
+        found_keys = {}
+        duplicated_keys = []
+        File.read(file).lines.each_with_index do |line, index|
+          key = extract_key(line: line)
+
+          next if key.nil?
+
+          # Check if that key was already encountered in the past
+          existing = found_keys[key]
+
+          duplicated_keys.append(key) unless existing.nil?
+          # TODO: Use proper Fastlane UI methods for this, or better rely on
+          # the consumer to read the return Hash (itself a TODO) and print an
+          # error. This is an helper, after all.
```

Comment:
@AliSoftware can you give me a 30% feedback on the direction I'm taking here?

Aside from the fact that I have the prod and test code in the same file (which is handy for me as I mess with this but obviously will change before calling this ready) what do you think of:

- Having this as a `Fastlane::Helper` – I don't see why it needs to belong to Fastlane, but all the helpers I looked at for inspiration do, so I've been consistent
- The idea of returning an Hash with key and an array of the line numbers where they occur.
- The idea of letting the caller look into the returned Hash and decided what to do with it, which right now should be something like emitting a warning via the `FastlaneCore::UI`

Thanks 🙇

Reply:
The fact that our helpers are in `Fastlane::Helper` namespace is likely a bit of a mix between legacy, and I'm assuming what the template for whatever `fastlane plugin create` command (or whatever it's called) might create as a scaffold when creating a new plugin?

It has always a bit bothered me tbh because that means we might accidentally clash with present or future helpers with the same name that fastlane itself would decide to use. I'd agree in refactoring all those to be in a `module WPMRT` or `module A8C` namespace instead at some point, if that makes more sense to all of us. This is part of the very long list of little things I'd love to refactor in our release-toolkit one day to clean things up and make it more structured, but never got the time to…

---

As for the return value and error emitting:
 - I think it makes more sense for the error emitting to be handled by either the action, or the client caller in the Fastfile, rather than the Helper. It makes more sense for the Helper to either `raise` or return a list of errors (like you did), and let the caller process them, because a Helper is more in the "business model" category than the "UI/View" category anyway
 - That being said, I like when the action at least offers an easy way to do the calls to `UI.message`/ `UI.user_error!`/… itself — as opposed to having the `Fastfile` calling the action be responsible for it. Because error handling and processing is quite repetitive and often the same, and it's easy to forget to handle it at call site if it's optional.

So my approach for that is usually:
 - Make the Helper just return the list or errors (as you did)
 - Make the action default to do the error handle that it received back from the helper, i.e. call `UI.user_error!` or `UI.message` and the like as necessary
 - Optionally, add a `ConfigItem` to the action to disable the error raising handled by the action, or choose a mode (e.g. between `:error`, `:warn`, `:silent`) in the rare case where the caller in `Fastfile` would prefer to handle the error reporting themselves.

That way we would have a sensible default that would avoid us to let those errors go unnoticed/unreported accidentally (for anyone just discovering that action and not realizing they'd have to handle the errors themselves), but an opt-out option for when you want more control.